src/V3DfgPeephole.cpp

```diff
@@ -180,54 +185,71 @@ class V3DfgPeephole final : public DfgVisitor {
         // the work list), but it will be deleted when the work list is processed.
         if (vtxp->getUser<DfgVertex*>()) return;
         // Otherwise we can delete it now.
+        // Remove from cache
+        m_cache.invalidateByValue(vtxp);
+        // Unlink source edges
+        vtxp->forEachSourceEdge([](DfgEdge& edge, size_t) { edge.unlinkSource(); });
+        // Should not have sinks
+        UASSERT_OBJ(!vtxp->hasSinks(), vtxp, "Should not delete used vertex");
+        //
         VL_DO_DANGLING(vtxp->unlinkDelete(m_dfg), vtxp);
     }
 
     void replace(DfgVertex* vtxp, DfgVertex* replacementp) {
+        UASSERT_OBJ(vtxp->width() == replacementp->width(), vtxp, "AAAA");
+        UASSERT_OBJ(vtxp != replacementp, vtxp, "WTF?");
         // Add sinks of replaced vertex to the work list
         addSinksToWorkList(vtxp);
         // Add replacement to the work list
         addToWorkList(replacementp);
         // Replace vertex with the replacement
+        vtxp->forEachSink([&](DfgVertex& sink) { m_cache.invalidateByValue(&sink); });
         vtxp->replaceWith(replacementp);
+        replacementp->forEachSink([&](DfgVertex& sink) { m_cache.cache(&sink); });
         // Vertex is now unused, so delete it
         deleteVertex(vtxp);
     }
 
-    void modified(DfgVertex* vtxp) {
-        // Add sinks of modified vertex to the work list
-        addSinksToWorkList(vtxp);
-        // Add the modified vertex itself to the work list
-        addToWorkList(vtxp);
-    }
-
     // Shorthand
     static AstNodeDType* dtypeForWidth(uint32_t width) { return DfgVertex::dtypeForWidth(width); }
 
     // Create a 32-bit DfgConst vertex
-    DfgConst* makeI32(FileLine* flp, uint32_t val) { return new DfgConst{m_dfg, flp, 32, val}; }
+    VL_ATTR_WARN_UNUSED_RESULT DfgConst* makeI32(FileLine* flp, uint32_t val) {
```

Comment:
I'm ok with your marking these as warn unused, but why do these specifically merit such special attention?   Usually that warning is used only for security problem flagging (not for e.g. leak prevention).

Reply:
Actually I added he attribute to the `bool` returning transformers below which caught me out several times, then maybe overdid it.